### PR TITLE
Graveler pre-commit/merge hooks

### LINKED ISSUE #1431: Graveler to trigger hooks during commit/merge

### PR DESCRIPTION
**Pull request by nopcoder**
closes #1431 
- Extend Graveler pre-commit and pre-merge hooks - calling a list of callbacks that on error will stop the commit/merge
- Model 'Action' document. An Action includes event and branch spec on which a list of hooks will be executed. A successful run of all hooks will enable the flow of commit/merge.

### CODE REVIEW

File: graveler/graveler.go
Comment (ozkatz) [inline graveler/graveler.go:317]: I'd accept a context here for logging and cancellation as well as key this by a repository ID as hooks belong to a specific repo

File: graveler/graveler.go
Comment (ozkatz) [inline graveler/graveler.go:899]: I think we should have an `ErrAbortedByHook` (bad name, but you get the idea) that will be propagated when a commit or merge operation fail. This will allow the API to return an informative error (possibly with a link to the failing hook for details, etc)

File: graveler/graveler.go
Comment (ozkatz) [inline graveler/graveler.go:1363]: assuming each hook is an independent unit that doesn't depend on previous hooks, not sure we want to run them serially.. I'm usually opposed to doing concurrency where it's not required, but I'm not sure it's not required :) we already provide a pretty narrow timeframe for hooks to complete in - this makes the situation worse by having all hooks compete for a slice of that time..

**Review by ozkatz — CHANGES_REQUESTED** [review]
I like how this fits pretty cleanly into Graveler! A testiment to the design. Overall looks good, proper error propagation and parallel execution are required imo. 

Also, not sure if a part of the PR - auditing of hook events (including them waiting to execute or are currently executing): will this be external to Graveler so in a different PR?

File: graveler/graveler.go
Comment (nopcoder) [reply graveler/graveler.go:317]: Graveler supplies the basic callback functionality we had in the cataloger.
There is no execution here, just getter.

File: graveler/graveler.go
Comment (nopcoder) [reply graveler/graveler.go:899]: True, as EntryCatalog will implement the actual load/parse/evaluate/execute of the actions it will use a specific error code with the execute/run ID to identify the specific call.

File: catalog/webhooks.go
Comment (itaiad200) [inline catalog/webhooks.go]: Maybe pass it to a dedicated package and avoid 'Hook' prefix everywhere?

File: graveler/graveler.go
Comment (nopcoder) [reply graveler/graveler.go:1363]: At graveler level, I don't have a problem running the hook one by one and keep it simple at this point.
Currently, we have only one consumer - entry catalog, and the execution of the webhooks will be in parallel.

File: graveler/graveler.go
Comment (itaiad200) [inline graveler/graveler.go:1245]: Sometimes merge fails due to conflict.

IMO it's a bad ux to run (what could be) a long running hook, let it pass, and then fail it immediately due to conflicts.

Is there a way to identify conflicts before calling the hook?

File: catalog/webhooks.go
Comment (nopcoder) [reply catalog/webhooks.go]: Will consolidate into 3 types - I don't think we need all of them - just generated them.

File: graveler/graveler.go
Comment (nopcoder) [reply graveler/graveler.go:1245]: sorry no - identify conflict will require going over the changes, which is a merge functionality.
if we hook implementation is consider small/fast at this point - I think it should stay between input parameter verification and the actual work.

Comment (itaiad200) [reply graveler/graveler.go:1245]: Not blocking on this on this PR, but this is something we need to take care of.

Think about the experience here: A user can try to make the hooks pass over and over again, pushing new code to his branch, committing and trying to merge. Finally, after several iterations the hook passed, but then he finds out immediately that all his work was for nothing - there are conflicts with the destination branch, he should of been fixing them first but lakeFS returned errors in the wrong order.

Comment (nopcoder) [reply graveler/graveler.go:1245]: UX point of view I agree, I would like any UI to tell me it is going to fail before I do any work.
It is something we need to discuss if moving the pre-merge step after the actual merge, but keep in mind that the pre-hook is there to verify the work that was done on the source ref, committed data - in the UI case can we can easily trigger 'diff' operation to identify conflict before submitting any merge.

File: graveler/graveler.go
Comment (nopcoder) [reply graveler/graveler.go:1363]: Moving the handling to EntryLevel and keeping simple callback at graveler level.

File: graveler/graveler.go
Comment (nopcoder) [reply graveler/graveler.go:899]: Used multi error package to format the final error.
It can be used with errors.Is and ErrAbortedByHook or the hook's error if needed.

File: graveler/graveler.go
Comment (nopcoder) [reply graveler/graveler.go:1245]: Moving the hook after the merge operation.

**Comment by nopcoder** [comment]
> Also, not sure if a part of the PR - auditing of hook events (including them waiting to execute or are currently executing): will this be external to Graveler so in a different PR?

Different PRs

File: catalog/action.go
Comment (itaiad200) [inline catalog/action.go:35]:
```suggestion
	ids := make(map[string]bool)
```
will give you a nicer usage:
```go
		if ids[hook.ID] {
			return fmt.Errorf("%w hook[%d] duplicate ID", ErrInvalidAction, i)
		}
```

File: catalog/action.go
Comment (itaiad200) [inline catalog/action.go:38]:
```suggestion
			return fmt.Errorf("hook[%d] missing ID: %w", i, ErrInvalidAction)
```

File: catalog/action.go
Comment (itaiad200) [inline catalog/action.go:41]: It's nicer to present the hook.ID (which is ok at this point):
```suggestion
			return fmt.Errorf("hook[%s] duplicate ID: %w", hook.ID, ErrInvalidAction)
```

File: catalog/action.go
Comment (itaiad200) [inline catalog/action.go:45]: Same here.

File: graveler/graveler.go
Comment (itaiad200) [inline graveler/graveler.go:1382]: This works for almost everything, unless original error was of type `*multierror.Error`

File: graveler/graveler.go
Comment (itaiad200) [inline graveler/graveler.go:1341]: Why do we need the getters?

File: graveler/graveler_test.go
Comment (itaiad200) [inline graveler/graveler_test.go:533]:
```suggestion
				t.Fatalf("Commit err=%v, expected ErrAbortedByHook", err)
```

**Review by itaiad200 — CHANGES_REQUESTED** [review]

File: catalog/action.go
Comment (nopcoder) [reply catalog/action.go:32]: 1. The verification here is for the existence of the event type
2. Not part of the verification - but empty will allow you to trigger the action on all branches.

File: catalog/action.go
Comment (nopcoder) [reply catalog/action.go:35]: I prefer the zero size value

File: graveler/graveler.go
Comment (nopcoder) [reply graveler/graveler.go:1341]: added just for the symmetry

**Review by itaiad200 — APPROVED** [review]
LGTM!

File: catalog/action.go
Comment (ozkatz) [inline catalog/action.go:29]: why not `-`?

**Review by ozkatz — APPROVED** [review]
Looks great!

**Pull request merged**
